src/lib/y2storage/autoinst_proposal.rb

```diff
@@ -0,0 +1,184 @@
+require "y2storage/exceptions"
+
+module Y2Storage
+  # Class to calculate a storage proposal for autoinstallation
+  #
+  # @example Example
+  #   profile = Yast::Profile.current["partitioning"]
+  #   proposal = Y2Storage::AutoinstProposal.new(partitioning)
+  #   proposal.proposed?            # => false
+  #   proposal.proposed_devicegraph # => nil
+  #   proposal.planned_devices      # => nil
+  #
+  #   proposal.propose              # => Performs the calculation
+  #
+  #   proposal.proposed?            # => true
+  #   proposal.proposed_devicegraph # => Proposed layout
+  class AutoinstProposal
+    include Yast::Logger
+
+    # @return [Hash] Partitioning layout from an AutoYaST profile
+    attr_reader :partitioning
+
+    # @return [Devicegraph] Initial device graph
+    attr_reader :initial_devicegraph
+
+    # Proposed layout of devices, nil if the proposal has not been
+    # calculated yet
+    # @return [Devicegraph]
+    attr_reader :proposed_devicegraph
+
+    alias_method :devices, :proposed_devicegraph
+
+    # Constructor
+    #
+    # @param partitioning [Array<Hash>] Partitioning schema from an AutoYaST profile
+    # @param devicegraph  [Devicegraph] starting point. If nil, then probed devicegraph
+    #   will be used
+    # @param disk_analyzer [DiskAnalyzer] if nil, a new one will be created
+    #   based in the initial devicegraph
+    def initialize(partitioning: [], devicegraph: nil, disk_analyzer: nil)
+      @partitioning = partitioning
+      @initial_devicegraph = devicegraph
+      @disk_analyzer = disk_analyzer
+      @proposed = false
+    end
+
+    # Checks whether the proposal has already been calculated
+    #
+    # @return [Boolean]
+    def proposed?
+      @proposed
+    end
+
+    # Calculates the proposal
+    #
+    # @raise [UnexpectedCallError] if called more than once
+    # @raise [NoDiskSpaceError] if there is no enough space to perform the installation
+    def propose
+      raise UnexpectedCallError if proposed?
+
+      drives = Proposal::AutoinstDrivesMap.new(initial_devicegraph, partitioning)
+
+      space_maker = Proposal::AutoinstSpaceMaker.new(disk_analyzer)
+      devicegraph = space_maker.provide_space(initial_devicegraph, drives)
+
+      @proposed_devicegraph = propose_devicegraph(devicegraph, drives)
+      @proposed = true
+
+      nil
+    end
+
+  protected
+
+    # Proposes a devicegraph based on given drives map
+    #
+    # This method falls back to #proposed_guided_devicegraph when the device map
+    # does not contain any partition.
+    #
+    # @param devicegraph [Devicegraph]                 Starting point
+    # @param drives      [Proposal::AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Devicegraph] Devicegraph containing the planned devices
+    def propose_devicegraph(devicegraph, drives)
+      planned_devices = plan_devices(devicegraph, drives)
+      create_devices(devicegraph, planned_devices, drives.disk_names)
+    rescue Error
```

Comment:
hmm, sorry, but no. This is not correct for more reasons:

1. what actually is wanted here is rescue only if there are no partitions specified. Why simple not check drives, if there is any partition and do not abuse exception for such case?

2. it rescue all Errors and always fallback to guided setup. What if error is in autoyast profile? Simply ignoring error and use different proposal is very from from my POV, as result is something that user do not want.
What's more, Error is also e.g. NoDiskSpaceError, which means, that there is not enough free space and it use instead different formatting then valid user specified, which is bad if user later depends on it e.g. in some of its autoyast script, which then failed which strange error because autoyast do not do what is requested in profile.